### PR TITLE
[vm] Load the writeset from StateView before committing it to the executor

### PR DESCRIPTION
**Pull request by runtian-zhou**
…cutor

## Motivation

As requested by @msmouse , Executor will expects writesets to be loaded from `StateView` before being applied. This PR will add this loading logic to the VM's logic of processing writesets.

Also the `check_change_set` logic is removed as it's no longer needed.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/libra/libra/blob/master/CONTRIBUTING.md#pull-requests)?

Yes

### CODE REVIEW

File: language/libra-vm/src/libra_vm.rs [hunk 1]
```diff
@@ -470,20 +466,34 @@ impl LibraVM {
...
 
+    fn load_writeset(
+        &self,
+        remote_cache: &BlockDataCache<'_>,
+        write_set: &WriteSet,
+    ) -> VMResult<()> {
+        // Load writeset contents into cache. Executor expects each writeop to be read from
+        // StateView before being able to commit them.
+        for op in write_set.iter() {
+            remote_cache.get(ap)?;
```
Comment (sblackshear) [inline language/libra-vm/src/libra_vm.rs:477]: `ap` -> `op`? Or `op` -> `ap` above?

File: language/libra-vm/src/libra_vm.rs [hunk 2]
```diff
@@ -470,20 +466,34 @@ impl LibraVM {
         result
     }
 
+    fn load_writeset(
```
Comment (sblackshear) [inline language/libra-vm/src/libra_vm.rs:469]: I find this name confusing. It sounds like you are loading the write set from the storage layer, whereas what you're actually doing is (IIUC) reading each access path that the write set is going to update.

I think it would be clearer to inline the function and expand the comment  to clarify (e.g., explain that all Move executions satisfy the read-before-write property).

File: language/libra-vm/src/libra_vm.rs [hunk 3]
```diff
@@ -238,18 +238,14 @@ impl LibraVM {
     fn check_change_set(&self, change_set: &ChangeSet, state_view: &dyn StateView) -> VMResult<()> {
         // This function is only invoked by WaypointWriteSet for now. We don't enforce the same
         // check on TransactionPayload::WriteSet.
-        if state_view.is_genesis() {
```
Comment (sblackshear) [inline language/libra-vm/src/libra_vm.rs:241]: Is this change related?

Comment (msmouse) [reply language/libra-vm/src/libra_vm.rs:241]: well, this is for the db-bootstrapper to be able to run a writeset txn on an existing state. We'll need to update the Executor accordingly so that it's still impossible to do it with normal consensus based flow.

Also, I'm not sure if it's entirely reasonable to "ban" deletions in such transactions? (I mean we are unleashing the devil through this txn anyways.)

Comment (sblackshear) [reply language/libra-vm/src/libra_vm.rs:241]: Good point; the "no deletions" check only makes sense for genesis.

Comment (runtian-zhou) [reply language/libra-vm/src/libra_vm.rs:241]: Do we want to just get rid of this check entirely given that we need to use this mechanism other than genesis?

Comment (sblackshear) [reply language/libra-vm/src/libra_vm.rs:241]: Yes, I think so. But I'm still wondering if removing the `is_genesis` check is related to the stated purpose of the PR?

Comment (msmouse) [reply language/libra-vm/src/libra_vm.rs:241]: Yeah removing the check and adding the read are both my requests, but can be separate commits.

Comment (sblackshear) [reply language/libra-vm/src/libra_vm.rs:241]: No need to split; just wanted to make sure I understood the relationship--thanks!

**Comment by msmouse** [comment]
Thanks @runtian-zhou 
@bors-libra r+
